### ISSUE TITLE
ONTAP modules generate "Unauthorized" audit events

### ISSUE THREAD

**Issue by mamoep** [ansible-collections/netapp#69]
##### SUMMARY

Usage of the ONTAP modules produce many "Unauthorized" audit events, even though the playbooks run successfully.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

Tested with na_ontap_info and na_ontap_command. Other modules might be affected aswell.

##### ANSIBLE VERSION

```paste below
ansible 2.9.10
  config file = /ansible/private/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.5 (default, Mar 20 2020, 17:08:22) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]

```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT

RHEL 7.9
ONTAP 9.7P7
Collection 20.11

##### STEPS TO REPRODUCE

1. Create custom user with admin role in ONTAP.

```Vserver: clustername
                                                                 Second
User/Group                 Authentication                 Acct   Authentication
Name           Application Method        Role Name        Locked Method
-------------- ----------- ------------- ---------------- ------ --------------
customuser    console     password      admin            no     none
customuser    http        password      admin            no     none
customuser    ontapi      password      admin            no     none
customuser    ssh         password      admin            no     none
```

2. Execute the playbook with the created user.

3. Check audit log with "security audit log show"
```Mon Nov 09 13:26:25 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c649 :: clustername:ontapi :: 192.168.1.2:33948 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:25 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c64b :: clustername:ontapi :: 192.168.1.2:33956 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:25 2020  clustername-01  [kern_audit:info:2509] 8503e8000006c64c :: clustername:ontapi :: 192.168.1.2:33958 :: clustername:customuser :: <netapp xmlns="http://www.netapp.com/filer/admin" version="1.110" vfiler="clustername"><ems-autosupport-log><computer-name>Ansible</computer-name><event-id>12345</event-id><event-source>na_ontap_info</event-source><app-version>20.11.0</app-version><category>Information</category><event-description>setup</event-description><log-level>6</log-level><auto-support>false</auto-support></ems-autosupport-log></netapp> :: Pending:
Mon Nov 09 13:26:25 2020  clustername-01  [kern_audit:info:2509] 8503e8000006c64c :: clustername:ontapi :: 192.168.1.2:33958 :: clustername:customuser :: ems-autosupport-log :: Success:
Mon Nov 09 13:26:25 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c64d :: clustername:ontapi :: 192.168.1.2:33960 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:25 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c64f :: clustername:ontapi :: 192.168.1.2:33964 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:26 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c651 :: clustername:ontapi :: 192.168.1.2:33968 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:26 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c653 :: clustername:ontapi :: 192.168.1.2:33976 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:26 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c655 :: clustername:ontapi :: 192.168.1.2:33982 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:26 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c657 :: clustername:ontapi :: 192.168.1.2:33988 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:26 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c659 :: clustername:ontapi :: 192.168.1.2:33994 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:26 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c65b :: clustername:ontapi :: 192.168.1.2:33998 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:26 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c65d :: clustername:ontapi :: 192.168.1.2:34008 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:26 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c65f :: clustername:ontapi :: 192.168.1.2:34012 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:26 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c661 :: clustername:ontapi :: 192.168.1.2:34018 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:26 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c663 :: clustername:ontapi :: 192.168.1.2:34022 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:27 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c665 :: clustername:ontapi :: 192.168.1.2:34030 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:27 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c667 :: clustername:ontapi :: 192.168.1.2:34034 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:27 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c669 :: clustername:ontapi :: 192.168.1.2:34040 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:27 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c66b :: clustername:ontapi :: 192.168.1.2:34046 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
Mon Nov 09 13:26:27 2020  clustername-01  [kern_audit:info:7967] 8503e8000006c66d :: clustername:ontapi :: 192.168.1.2:34050 :: clustername:unknown :: POST /servlets/netapp.servlets.admin.XMLrequest_filer HTTP/1.1 :: Error: 401 Unauthorized
```

```yaml
- hosts: localhost
  become: false
  collections:
    - netapp.ontap
  name: ONTAP facts gatherer
  vars:
    out_path: /tmp
    username: customuser
    password:  pass
  tasks:
  - name: Gather facts
    na_ontap_info:
      hostname: "clustername.dzbank.vrnet"
      username: "{{ username }}"
      password: "{{ password }}"
      https: true
      validate_certs: true
      use_rest: always
      gather_subset:
        - "cifs_server_info"
        - "cifs_vserver_security_info"
        - "cluster_identity_info"
        - "ldap_client"
        - "ldap_config"
        - "net_dev_discovery_info"
        - "net_dns_info"
        - "net_ifgrp_info"
        - "net_port_info"
        - "ntp_server_info"
        - "security_login_account_info"
        - "sis_info"
        - "vserver_info"
    register: config_out
```

##### EXPECTED RESULTS

Successful login in the audit log.

##### ACTUAL RESULTS

Many failed logins in the audit log.

```paste below

```

**Comment by lonico** [comment]
It's interesting these are POST events, and under clustername:unknown.  The only POST event should be the EMS log, and it went through, under clustername:customuser.

Thank you for noticing and reporting this.   An interesting mystery.

**Comment by lonico** [comment]
I think this is the way Basic Authentication is supposed to work.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication

It seems we could take a shortcut, but this is the way urllib is working (netapp-lib is using urllib under the cover).  It first sends an request without authorization, waits for a 401, and resends the request with an Authorization header.  I did not find a way to change the behavior.

(requests seems to do a better job)

**Comment by mamoep** [comment]
According to Stack Overflow you could create the Authorization header yourself and add it to the request. Did you mean this by shortcut?
Python2: https://stackoverflow.com/a/44239923
Python3: https://stackoverflow.com/a/24648149/4609258

Our security team is monitoring the audit logs and is not amused of this amount of "failed logins". It would be really helpful if you could fix that.

**Comment by lonico** [comment]
Yes, this is what I was thinking of doing, but I would have liked something more built-in rather than manually constructing a header.
But this requires a change in netapp-lib, and I'm sure how wise it is to deviate from the RFC.  We'll look more into it.

Maybe a better alternative is to use SSL certificates.  I checked this morning, and I don't see any 401 error.   And it has the advantage over basic authentication that passwords are not sent in (almost) clear.

**Comment by lonico** [comment]
working on it

**Comment by lonico** [comment]
fixed with 860d4b2e8d74f7504273a2526e48405fa0654598
This will be released in Galaxy as 20.12.0 early December.

**Comment by mamoep** [comment]
Thank you very much for the fast fix!
